sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/StaxXmlGenerator.scala

```diff
@@ -241,4 +252,165 @@ class StaxXmlGenerator(
+            // an attribute or not.
+            // In addition, we also check if the field is a value tag, in case the value tag also
+            // starts with the attribute prefix.
+            f.startsWith(options.attributePrefix) && f != options.valueTag
+        }
+
+        // We need to write attributes first before the elements.
+        (attributes ++ elements).foreach {
+          case (field, elementPos) =>
+            writeVariant(field, v, elementPos)
+        }
+      }
+    )
+    gen.writeEndElement()
+  }
+
+  /**
+   * Write a Variant array to XML. A Variant array is flattened and written as a sequence of
+   * XML element with the same element name as the array field name.
+   *
+   * @param name The name of the array field
+   * @param v The original Variant entity
+   * @param pos The position in the Variant data array where the array value starts
+   */
+  private def writeVariantArray(name: String, v: VariantVal, pos: Int): Unit = {
+    handleArray(
+      v.getValue,
+      pos,
+      (size, offsetSize, offsetStart, dataStart) => {
+        // Traverse each item of the array and write each of them as an XML element
+        (0 until size).foreach { i =>
+          val offset =
+            readUnsigned(v.getValue, offsetStart + offsetSize * i, offsetSize)
+          val elementPos = dataStart + offset
+          // Check if the array element is also of type ARRAY
+          if (getType(v.getValue, elementPos) == Type.ARRAY) {
+            // For the case round trip in reading and writing XML files, [[ArrayType]] cannot have
+            // [[ArrayType]] as element type. It always wraps the element with [[StructType]]. So,
+            // this case only can happen when we convert a normal [[DataFrame]] to XML file.
+            // When [[ArrayType]] has [[ArrayType]] as elements, it is confusing what is element
+            // name for XML file.
+            writeVariantArray(options.arrayElementName, v, elementPos)
+          } else {
+            writeVariant(name, v, elementPos)
+          }
+        }
+      }
+    )
+  }
+
+  /**
+   * Write a Variant primitive field to XML
+   *
+   * @param name The name of the field
+   * @param v The original Variant entity
+   * @param pos The position in the Variant data array where the field value starts
+   */
+  private def writeVariantPrimitive(name: String, v: VariantVal, pos: Int): Unit = {
+    val primitiveVal = getType(v.getValue, pos) match {
+      case Type.NULL => Option(options.nullValue).orNull
+      case Type.BOOLEAN =>
+        getBoolean(v.getValue, pos)
+      case Type.LONG =>
+        getLong(v.getValue, pos)
+      case Type.STRING =>
+        getString(v.getValue, pos)
+      case Type.DOUBLE =>
+        getDouble(v.getValue, pos)
+      case Type.DECIMAL =>
+        getDecimal(v.getValue, pos)
+      case Type.DATE =>
+        dateFormatter.format(LocalDate.ofEpochDay(getLong(v.getValue, pos)))
```

Comment:
Is it necessary to convert to LocalDate?
```suggestion
        dateFormatter.format(getLong(v.getValue, pos))
```

Reply:
verified that it's not needed.